### PR TITLE
[HotFix] Resolve TPU Training

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #6004 #5841

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: dockers/tpu-tests/tpu_test_cases.jsonnet [hunk 1]
```diff
@@ -25,6 +25,12 @@ local tputests = base.BaseTest {
           pytorch_lightning/utilities/xla_device_utils.py \
           tests/accelerators/test_tpu_backend.py \
           tests/models/test_tpu.py
+      # Takes too long
+      # python pl_examples/domain_templates/computer_vision_fine_tuning.py \
+      #  --tpu_cores 8 \
+      #  --epochs 15 \
+      #  --limit_train_batches 8 \
+      #  --limit_val_batches 8
```
Comment (carmocca) [inline dockers/tpu-tests/tpu_test_cases.jsonnet:33]: What is the point of adding these commented lines?

File: pl_examples/domain_templates/computer_vision_fine_tuning.py [hunk 2]
```diff
@@ -71,7 +71,7 @@ def __init__(self, milestones: tuple = (5, 10), train_bn: bool = False):
         self.train_bn = train_bn
 
     def freeze_before_training(self, pl_module: pl.LightningModule):
-        self.freeze(modules=pl_module.feature_extractor, train_bn=self.train_bn)
+        self.freeze(modules=pl_module.feature_extractor, train_bn=False)
```
Comment (carmocca) [inline pl_examples/domain_templates/computer_vision_fine_tuning.py:74]: Why this change? Is it related to the fix?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -554,6 +554,12 @@ def _save_top_k_checkpoints(self, trainer, pl_module, metrics):
         epoch = metrics.get("epoch")
         step = metrics.get("step")
 
+        # when `val_loss` is being logged and no ModelCheckpoint is being provided
+        # `val_loss` will be selected for monitor and need to be reduced to
+        # prevent processes divergence
+        if self.monitor == "val_loss":
+            current = trainer.training_type_plugin.reduce(current, reduce_op="mean")
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:561]: Can you address this comment? From the other PR

https://github.com/PyTorchLightning/pytorch-lightning/pull/6004#discussion_r576778261

Even though this might fix the bug, it really doesn't seem like the place to  add this fix as it pollutes the ModelCheckpoint with unrelated logic and violates the call-hierarchy recently defined in the accelerator refactor.

cc: @awaelchli @justusschock

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -213,6 +217,10 @@ def __load_weights_on_main_process(self) -> None:
 
         self._model = model
 
+    def _close_logger(self, trainer) -> None:
+        if hasattr(trainer, "logger"):
+            trainer.logger.close()
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:222]: Not all loggers define `close`, this will fail if not tensorboard. Why not just use `finalize`?

File: pytorch_lightning/loggers/tensorboard.py [hunk 5]
```diff
@@ -234,6 +234,9 @@ def save(self) -> None:
 
     @rank_zero_only
     def finalize(self, status: str) -> None:
+        self.close()
+
+    def close(self):
```
Comment (carmocca) [inline pytorch_lightning/loggers/tensorboard.py:239]: If this is not part of the public logger API, you should prepend with `_`

File: pl_examples/domain_templates/computer_vision_fine_tuning.py [hunk 6]
```diff
@@ -55,7 +55,7 @@
 import pytorch_lightning as pl
 from pl_examples import cli_lightning_logo
 from pytorch_lightning import _logger as log
-from pytorch_lightning import LightningDataModule
+from pytorch_lightning import LightningDataModule, seed_everything
```
Comment (justusschock) [inline pl_examples/domain_templates/computer_vision_fine_tuning.py:58]: all changes in this file should be irrelevant I think.

File: pytorch_lightning/loggers/tensorboard.py (hunk 5 above)
Comment (justusschock) [reply pytorch_lightning/loggers/tensorboard.py:239]: it should be removed completely and finalize should be used

File: tests/helpers/utils.py [hunk 7]
```diff
@@ -91,7 +91,7 @@ def wrapper(*args, **kwargs):
 
         def inner_f(queue, **kwargs):
             try:
-                func(**kwargs)
+                func(*args, **kwargs)
```
Comment (justusschock) [inline tests/helpers/utils.py:94]: This uses kwargs from` inner_f` and args from `wrapper`. Pretty sure this is not intended like this

**Review by justusschock — COMMENTED** [review]
fully agree with @carmocca review + some additional points

File: dockers/tpu-tests/tpu_test_cases.jsonnet (hunk 1 above)
Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet:33]: I was thinking to activate it later, but still take too long.

Comment (tchaton) [reply dockers/tpu-tests/tpu_test_cases.jsonnet:33]: Removed !

File: pl_examples/domain_templates/computer_vision_fine_tuning.py (hunk 2 above)
Comment (tchaton) [reply pl_examples/domain_templates/computer_vision_fine_tuning.py:74]: Just to make training faster, but still slow.

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:561]: I closed the other PR.

File: CHANGELOG.md [hunk 8]
```diff
@@ -288,6 +288,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed missing `process_dataloader` call for `TPUSpawn` when in distributed mode ([#6015](https://github.com/PyTorchLightning/pytorch-lightning/pull/6015))
 
 
+- Fixed synchrnization issues with TPUs Training ([#6027](https://github.com/PyTorchLightning/pytorch-lightning/pull/6027))
```
Comment (carmocca) [inline CHANGELOG.md:291]:
```suggestion
- Fixed synchronization issues with TPU training ([#6027](https://github.com/PyTorchLightning/pytorch-lightning/pull/6027))
```

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:561]: Would be moved in another PR.

File: pytorch_lightning/loggers/tensorboard.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/loggers/tensorboard.py:239]: Move back !

File: tests/helpers/utils.py (hunk 7 above)
Comment (tchaton) [reply tests/helpers/utils.py:94]: Not sure to understand why it shouldn't passed. But happy to revert it back. Worked this fine :)

File: CHANGELOG.md (hunk 8 above)
Comment (tchaton) [reply CHANGELOG.md:291]: Done !

File: tests/helpers/utils.py (hunk 7 above)
Comment (tchaton) [reply tests/helpers/utils.py:94]: Done !

File: pl_examples/domain_templates/computer_vision_fine_tuning.py (hunk 6 above)
Comment (tchaton) [reply pl_examples/domain_templates/computer_vision_fine_tuning.py:58]: Zach is using this script for making sure TPU works. Just added everything to make it compatible with TPU training.

File: pl_examples/domain_templates/computer_vision_fine_tuning.py (hunk 2 above)
Comment (carmocca) [reply pl_examples/domain_templates/computer_vision_fine_tuning.py]: train_bn is set to False by default, but the argument wasn't correctly passed. Fixed in https://github.com/PyTorchLightning/pytorch-lightning/pull/6027/commits/efcff122e7c8560c1699dbb82426a433f7a75527

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 9]
```diff
@@ -225,9 +233,11 @@ def start_training(self, trainer) -> None:
         # todo: precision pluging is call in accelerator setup and should be moved
         if 'XLA_USE_BF16' in os.environ:
             del os.environ["XLA_USE_BF16"]
+        self._close_logger(trainer)
         xmp.spawn(self.new_process, **self.xmp_spawn_kwargs)
 
     def start_testing(self, trainer) -> None:
+        self._close_logger(trainer)
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:240]: Why do we do this?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 10]
```diff
@@ -554,6 +554,13 @@ def _save_top_k_checkpoints(self, trainer, pl_module, metrics):
         epoch = metrics.get("epoch")
         step = metrics.get("step")
 
+        # when `val_loss` is being logged and no ModelCheckpoint is being provided
+        # `val_loss` will be selected for monitor and need to be reduced to
+        # prevent processes divergence
+        # Todo: Move this logic to logger_connector
```
Comment (carmocca) [inline pytorch_lightning/callbacks/model_checkpoint.py:560]:
```suggestion
        # TODO: Move this logic to logger_connector. This also needs to be fixed for any other logged value
        # which will be monitored by ModelCheckpoint and is not logged with `sync_dist=True`
```

File: pytorch_lightning/trainer/training_loop.py [hunk 11]
```diff
@@ -140,7 +140,7 @@ def on_train_end(self):
         # todo: TPU 8 cores hangs in flush with TensorBoard. Might do for all loggers.
         # It might be related to xla tensors blocked when moving the cpu
         # kill loggers
-        if self.trainer.logger is not None and self.trainer.training_type_plugin.should_finalize:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:143]: aha!

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 12]
```diff
@@ -179,6 +175,14 @@ def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
...
         return should_stop
 
+    def reduce(self, output, group: Optional[Any] = None, reduce_op: str = None):
+        if not isinstance(output, torch.Tensor):
+            output = torch.tensor(output, device=self.device)
+        output = xm.mesh_reduce('reduce', output, sum)
+        if isinstance(reduce_op, str) and reduce_op.lower() == "mean":
+            output /= self.world_size
+        return output
+
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:185]: nice, I'll add docs etc in #6011

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:240]: @tchaton is it a pickling issue?

File: pl_examples/domain_templates/computer_vision_fine_tuning.py [hunk 13]
```diff
@@ -276,10 +274,23 @@ def add_model_specific_args(parent_parser):
...
+            "--limit_train_batches",
+            default=1.0,
+            type=float,
+            help="How much of training dataset to check (floats = percent, int = num_batches)"
+        )
+        parser.add_argument(
+            "--limit_val_batches",
+            default=1.0,
+            type=float,
+            help="How much of validation dataset to check (floats = percent, int = num_batches)"
```
Comment (awaelchli) [inline pl_examples/domain_templates/computer_vision_fine_tuning.py:289]: the trainer already adds these arguments, there is a collision now as they cannot exist twice

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:240]: The main process logger applies a lock on the log_dir. I need to close the logged, so the spawned rank zero process can actually take the control back and start logging. If not, the process will be locked.

File: pytorch_lightning/trainer/training_loop.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:143]: Yes, closing the logger solves the issue.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 12 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:185]: Great !

File: pl_examples/domain_templates/computer_vision_fine_tuning.py (hunk 13 above)
Comment (tchaton) [reply pl_examples/domain_templates/computer_vision_fine_tuning.py:289]: The trainer isn't used with the parser.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:240]: And it needs to be done for both train, test as users might not always call fit, but test directly with multiple tpu cores.

File: pl_examples/domain_templates/computer_vision_fine_tuning.py [hunk 14]
```diff
@@ -276,10 +274,23 @@ def add_model_specific_args(parent_parser):
             help="Name (as in ``torchvision.models``) of the feature extractor",
         )
         parser.add_argument(
-            "--epochs", default=15, type=int, metavar="N", help="total number of epochs", dest="nb_epochs"
+            "--epochs", default=5, type=int, metavar="N", help="total number of epochs", dest="nb_epochs"
```
Comment (Borda) [inline pl_examples/domain_templates/computer_vision_fine_tuning.py:277]: why not `max_epochs` from Trainer?

File: pytorch_lightning/accelerators/accelerator.py [hunk 15]
```diff
@@ -388,3 +389,11 @@ def all_gather(self, tensor: Union[torch.Tensor], group: Optional[Any] = None, s
             A tensor of shape (world_size, batch, ...)
         """
         return all_gather_ddp_if_available(tensor, group=group, sync_grads=sync_grads)
+
+    def process_dataloader(self, dataloader: Union[Iterable, DataLoader]) -> Union[Iterable, DataLoader]:
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:393]: now it is duplicate to Plugin's `process_dataloader`?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 16]
```diff
@@ -179,6 +175,14 @@ def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
         should_stop = int(stop.item()) == self.world_size
         return should_stop
 
+    def reduce(self, output, group: Optional[Any] = None, reduce_op: str = None):
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:178]: this effectively means that only sum or mean is valid `reduce_op` right?

File: pytorch_lightning/trainer/training_loop.py (hunk 11 above)
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:143]: I remember that this was added for a reason, mind check the PR which added it?

Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:143]: I added this in my previous TPU PR. This PR resolve the bug `should_finalize` prevented.

File: pl_examples/domain_templates/computer_vision_fine_tuning.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pl_examples/domain_templates/computer_vision_fine_tuning.py:289]: Removed every change to this file.

File: pl_examples/domain_templates/computer_vision_fine_tuning.py (hunk 14 above)
Comment (tchaton) [reply pl_examples/domain_templates/computer_vision_fine_tuning.py:277]: Removed every change to this file.

File: pytorch_lightning/accelerators/accelerator.py (hunk 15 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:393]: It is the intended behaviour. Ideally, the trainer should interact only with the accelerator, so `process_dataloader` from accelerator redirect to training_type_plugin.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py]: For now, yes.

Comment (Borda) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py]: so let's add there assert 
```
assert reduce_op in (None, 'sum', 'mean')
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 17]
```diff
@@ -179,6 +176,29 @@ def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
         should_stop = int(stop.item()) == self.world_size
         return should_stop
 
+    def reduce(self, output, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = None):
+        if not isinstance(output, torch.Tensor):
+            output = torch.tensor(output, device=self.device)
+
+        if (isinstance(reduce_op, ReduceOp) and ReduceOp != ReduceOp.SUM) \
+           or isinstance(reduce_op, str) and reduce_op.lower() not in ("sum", "mean", "avg"):
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:184]:
```suggestion
        _valid_reduce_op_str = isinstance(reduce_op, str) and reduce_op.lower() not in ("sum", "mean", "avg")
        if (isinstance(reduce_op, ReduceOp) and reduce_op != ReduceOp.SUM) or _valid_reduce_op_str:
```
but what is this one `ReduceOp != ReduceOp.SUM`? shall it be `reduce_op != ReduceOp.SUM`

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 18]
```diff
@@ -179,6 +176,29 @@ def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
...
+        divide_by_world_size = False
+
+        output = xm.mesh_reduce('reduce', output, sum)
+
+        if isinstance(reduce_op, str) and reduce_op.lower() in ("avg", "mean"):
+            divide_by_world_size = True
+        # sync all processes before reduction
+
+        if divide_by_world_size:
+            output = output / self.world_size
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:198]:
```suggestion
        output = xm.mesh_reduce('reduce', output, sum)

        if isinstance(reduce_op, str) and reduce_op.lower() in ("avg", "mean"):
            output = output / self.world_size
```

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 19]
```diff
@@ -179,6 +176,23 @@ def reduce_early_stopping_decision(self, should_stop: bool) -> bool:
         should_stop = int(stop.item()) == self.world_size
         return should_stop
 
+    def reduce(self, output, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = None):
+        if not isinstance(output, torch.Tensor):
+            output = torch.tensor(output, device=self.device)
+
+        _valid_reduce_op_str = isinstance(reduce_op, str) and reduce_op.lower() not in ("sum", "mean", "avg")
+        if (isinstance(reduce_op, ReduceOp) and reduce_op != ReduceOp.SUM) or _valid_reduce_op_str:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:184]:
```suggestion
        _invalid_reduce_op = isinstance(reduce_op, ReduceOp) and reduce_op != ReduceOp.SUM
        _invalid_reduce_op_str = isinstance(reduce_op, str) and reduce_op.lower() not in ("sum", "mean", "avg")
        if _invalid_reduce_op or _invalid_reduce_op_str:
```
sorry for typo

**Review by Borda — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
